File: app/lib/domain/transcript/transcript_event.dart
import 'package:app/domain/session_state.dart';
import 'package:app/protocol/protocol.dart'
    show Usage, UserMessageStreamingBehavior;

/// Capture one immutable fact in a session's append-only transcript stream.
sealed class TranscriptEvent {
  const TranscriptEvent({
    required this.eventId,
    required this.sessionId,
    required this.ts,
    this.turnId,
  });

  final String eventId;
  final String sessionId;
  final DateTime ts;
  final String? turnId;
}

/// Record a locally accepted user submission before server confirmation.
final class UserMessageSubmitted extends TranscriptEvent {
  const UserMessageSubmitted({
    required super.eventId,
    required super.sessionId,
    required super.ts,
    super.turnId,
    required this.clientMessageId,
    required this.text,
    this.image,
    this.held = false,
    this.awaitingPickup = false,
  });

  final String clientMessageId;
  final String text;
  final MessageImage? image;

  /// Keep a steered submission out of transcript order until canonical pickup.
  final bool awaitingPickup;

  /// `true` when the first attempt was held locally because the room was
  /// offline; `false` when it reached a channel. This remains provenance for
  /// transcript/UI projection only. The encrypted owner-delivery outbox is the
  /// single resend authority for both held and sent-but-unconfirmed messages.
  final bool held;
}

/// Record the authoritative confirmation of a user message.
final class UserMessageConfirmed extends TranscriptEvent {
  const UserMessageConfirmed({
    required super.eventId,
    required super.sessionId,
    required super.ts,
    super.turnId,
    required this.clientMessageId,
    required this.text,
    this.image,
    this.streamingBehavior,
    this.semanticPickup = true,
  });

  final String clientMessageId;
  final String text;
  final MessageImage? image;
  final UserMessageStreamingBehavior? streamingBehavior;

  /// Distinguish delivery acceptance from the timestamped agent pickup event.
  ///
  /// Old persisted records default to `true`, preserving their prior order.
  final bool semanticPickup;
}

/// Record a terminal local failure while a submission lacks confirmation.
final class UserMessageFailed extends TranscriptEvent {
  const UserMessageFailed({
    required super.eventId,
    required super.sessionId,
    required super.ts,
    super.turnId,
    required this.clientMessageId,
    required this.code,
    required this.message,
  });

  final String clientMessageId;
  final String code;
  final String message;
}

/// Record one streamed assistant delta for an in-progress reply.
final class AssistantDeltaReceived extends TranscriptEvent {
  const AssistantDeltaReceived({
    required super.eventId,
    required super.sessionId,
    required super.ts,
    super.turnId,
    required this.replyTo,
    required this.delta,
  });

  final String replyTo;
  final String delta;
}

/// Record the authoritative assistant reply that replaces streamed deltas.
final class AssistantMessageCommitted extends TranscriptEvent {
  const AssistantMessageCommitted({
    required super.eventId,
    required super.sessionId,
    required super.ts,
    super.turnId,
    required this.messageId,
    required this.replyTo,
    required this.text,
    this.usage,
  });

  final String messageId;
  final String replyTo;
  final String text;
  final Usage? usage;
}

/// Record completion of an assistant turn that may have no final message.
final class AssistantDoneReceived extends TranscriptEvent {
  const AssistantDoneReceived({
    required super.eventId,
    required super.sessionId,
    required super.ts,
    super.turnId,
    required this.replyTo,
    this.usage,
  });

  final String replyTo;
  final Usage? usage;
}

/// Record a tool invocation requested during a turn.
final class ToolRequested extends TranscriptEvent {
  const ToolRequested({
    required super.eventId,
    required super.sessionId,
    required super.ts,
    super.turnId,
    required this.toolCallId,
    required this.tool,
    required this.args,
  });

  final String toolCallId;
  final String tool;
  final Map<String, Object?> args;
}

/// Record the result or failure that settles a requested tool invocation.
final class ToolFinished extends TranscriptEvent {
  const ToolFinished({
    required super.eventId,
    required super.sessionId,
    required super.ts,
    super.turnId,
    required this.toolCallId,
    this.result,
    this.error,
  });

  final String toolCallId;
  final Object? result;
  final String? error;
}

/// Record a context compaction and its user-visible recap.
final class CompactionRecorded extends TranscriptEvent {
  const CompactionRecorded({
    required super.eventId,
    required super.sessionId,
    required super.ts,
    super.turnId,
    required this.summary,
    this.tokensBefore,
  });

  final String summary;
  final int? tokensBefore;
}

File: pi-extension/src/session/transcript_event.ts
import type { StreamingBehavior, Usage, WireImage } from "../protocol/types.js";

export type TranscriptEvent =
  | {
      kind: "user_submitted";
      eventId: string;
      sessionId: string;
      ts: number;
      turnId?: string;
      clientMessageId: string;
      text: string;
      images?: WireImage[];
    }
  | {
      kind: "user_confirmed";
      eventId: string;
      sessionId: string;
      ts: number;
      turnId?: string;
      clientMessageId: string;
      text: string;
      images?: WireImage[];
      streamingBehavior?: StreamingBehavior;
    }
  | {
      kind: "user_failed";
      eventId: string;
      sessionId: string;
      ts: number;
      turnId?: string;
      clientMessageId: string;
      code: string;
      message: string;
    }
  | {
      kind: "assistant_delta";
      eventId: string;
      sessionId: string;
      ts: number;
      turnId?: string;
      replyTo: string;
      delta: string;
    }
  | {
      kind: "assistant_committed";
      eventId: string;
      sessionId: string;
      ts: number;
      turnId?: string;
      messageId: string;
      replyTo: string;
      text: string;
      usage?: Usage;
    }
  | {
      kind: "assistant_done";
      eventId: string;
      sessionId: string;
      ts: number;
      turnId?: string;
      replyTo: string;
      usage?: Usage;
    }
  | {
      kind: "provider_error";
      eventId: string;
      sessionId: string;
      ts: number;
      turnId?: string;
      replyTo?: string;
      code: string;
      message: string;
    }
  | {
      kind: "tool_requested";
      eventId: string;
      sessionId: string;
      ts: number;
      turnId?: string;
      toolCallId: string;
      tool: string;
      args: Record<string, unknown>;
    }
  | {
      kind: "tool_finished";
      eventId: string;
      sessionId: string;
      ts: number;
      turnId?: string;
      toolCallId: string;
      result?: unknown;
      error?: string;
    }
  | {
      kind: "compaction_recorded";
      eventId: string;
      sessionId: string;
      ts: number;
      turnId?: string;
      summary: string;
      tokensBefore?: number;
    };

export type TranscriptTurnStatus = "idle" | "working" | "streaming" | "error";

export interface TranscriptTurnView {
  status: TranscriptTurnStatus;
  replyTo?: string;
  error?: string;
}

export interface TranscriptProjection<TMessage = unknown> {
  messages: TMessage[];
  streaming?: { replyTo: string; text: string };
  turn: TranscriptTurnView;
}
